### PR TITLE
yarn: Fix handling of empty outdated & scoped packages

### PR DESCRIPTION
**Pull request by moonrail**
##### SUMMARY
This PR fixes two issues with the yarn module:
1. `Index out of range` with `state: 'latest'` and no outdated packages
2. `Too many values to unpack` with scoped packages installed

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
yarn

###### 1. `Index out of range` with `state: 'latest'` and no outdated packages
Example:
```yaml
- yarn:
    path: '/var/www/test.domain.tld/some_app'
    name: 'imagemin'
    state: 'latest'
```
```bash
Traceback (most recent call last):
  File "/var/tmp/ansible-tmp-1591514673.7097096-9690-36426478672085/AnsiballZ_yarn.py", line 102, in <module>
    _ansiballz_main()
  File "/var/tmp/ansible-tmp-1591514673.7097096-9690-36426478672085/AnsiballZ_yarn.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/var/tmp/ansible-tmp-1591514673.7097096-9690-36426478672085/AnsiballZ_yarn.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.packaging.language.yarn', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_yarn_payload_6g8x8vs_/ansible_yarn_payload.zip/ansible/modules/packaging/language/yarn.py", line 389, in <module>
  File "/tmp/ansible_yarn_payload_6g8x8vs_/ansible_yarn_payload.zip/ansible/modules/packaging/language/yarn.py", line 371, in main
  File "/tmp/ansible_yarn_payload_6g8x8vs_/ansible_yarn_payload.zip/ansible/modules/packaging/language/yarn.py", line 281, in list_outdated
IndexError: list index out of range
```
Situation on machine would be e.g.:
```bash
centos8:/var/www/test.domain.tld/some_app $ cat package.json 
{
  "name": "some_app",
  "license": "UNLICENSED",
  "scripts": [
    {
      "start": "/var/www/test.domain.tld/some_app/some_script.js"
    }
  ],
  "dependencies": {
    "imagemin": "^7.0.1",
    "package": "^1.0.1"
  }
}
centos8:/var/www/test.domain.tld/some_app $ yarn outdated --json
centos8:/var/www/test.domain.tld/some_app $ yarn -v
1.22.4
centos8:/var/www/test.domain.tld/some_app $
```

###### 2. `Too many values to unpack` with scoped packages installed
Example:
```yaml
- yarn:
    path: '/var/www/test.domain.tld/some_app'
    name: 'imagemin'
    state: 'latest'
```
```bash
Traceback (most recent call last):
  File "/var/tmp/ansible-tmp-1591518509.349181-16492-49670775886975/AnsiballZ_yarn.py", line 102, in <module>
    _ansiballz_main()
  File "/var/tmp/ansible-tmp-1591518509.349181-16492-49670775886975/AnsiballZ_yarn.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/var/tmp/ansible-tmp-1591518509.349181-16492-49670775886975/AnsiballZ_yarn.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.packaging.language.yarn', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_yarn_payload_jzqnrkyd/ansible_yarn_payload.zip/ansible/modules/packaging/language/yarn.py", line 392, in <module>
  File "/tmp/ansible_yarn_payload_jzqnrkyd/ansible_yarn_payload.zip/ansible/modules/packaging/language/yarn.py", line 373, in main
  File "/tmp/ansible_yarn_payload_jzqnrkyd/ansible_yarn_payload.zip/ansible/modules/packaging/language/yarn.py", line 250, in list
ValueError: too many values to unpack (expected 2)
```
Situation on machine would be e.g.:
```bash
centos8:/var/www/test.domain.tld/some_app $ yarn list --depth=0 --json | jq
{
  "type": "tree",
  "data": {
    "type": "list",
    "trees": [
      {
        "name": "imagemin@7.0.1",
        "children": [],
        "hint": null,
        "color": null,
        "depth": 0
      },
      (...)
      {
        "name": "@nodelib/fs.stat@2.0.3",
        "children": [],
        "hint": null,
        "color": null,
        "depth": 0
      },
      {
        "name": "@types/node@14.0.11",
        "children": [],
        "hint": null,
        "color": null,
        "depth": 0
      },
      (...)
    ]
  }
}
```

### CODE REVIEW

File: changelogs/fragments/474-yarn_fix-outdated-fix-list.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - yarn - Fixed an index out of range error when no outdated packages where returned by yarn executable (see https://github.com/ansible-collections/community.general/pull/474)
+  - yarn - Fixed an too many values to unpack error when scoped packages are installed (see https://github.com/ansible-collections/community.general/pull/474)
```
Comment (Andersson007) [inline changelogs/fragments/474-yarn_fix-outdated-fix-list.yml:3]:
```suggestion
  - yarn - fixed an index out of range error when no outdated packages where returned by yarn executable (see https://github.com/ansible-collections/community.general/pull/474).
  - yarn - fixed an too many values to unpack error when scoped packages are installed (see https://github.com/ansible-collections/community.general/pull/474).
```
see the latest doc https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to

**Comment by Andersson007** [comment]
we have ci tests for the module tests/integration/targets/yarn/tasks/
would be cool if you reproduce the correct handling of the cases there

**Comment by moonrail** [comment]
@Andersson007 
I've looked into [tests](https://github.com/ansible-collections/community.general/blob/aebc1b03fdc1a5e56f7e7bf34784f8b1d83174e2/tests/integration/targets/yarn/tasks/main.yml#L26) and found following versioning test cases:
- Node.js 4.8.0 & Yarn 1.6.0
- Node.js 8.0.0 & Yarn 1.6.0

As Yarn is currently at 1.22.4 and Node.js around 14.X, I propose another test case to be added:
- Node.js 12.18.0 ([LTS](https://nodejs.org/en/download/releases/)) & Yarn 1.22.4 ([current](https://github.com/yarnpkg/yarn/releases))

I do not know how to upload files to https://ansible-ci-files.s3.amazonaws.com/test/integration/targets/yarn/, but the versions mentioned above should be added there to be able to use them for testing, as they're downloaded from there:
https://github.com/ansible-collections/community.general/blob/aebc1b03fdc1a5e56f7e7bf34784f8b1d83174e2/tests/integration/targets/yarn/tasks/run.yml#L8

---

For testing scoping I've added a scoped package to package.j2 (package.json) thats being tested with. It will be installed and further handling of packages will be with an installed scoped package.

**Comment by Andersson007** [comment]
@moonrail sorry for late reply. Would you like to add packages to ansible aws and update the tests?

**Comment by Andersson007** [comment]
@verkaufer @chrishoffman could you please look at the changes?

**Comment by Andersson007** [comment]
I've just emailed the authors

**Comment by felixfontein** [comment]
I'm not sure who has access to https://ansible-ci-files.s3.amazonaws.com/. Maybe @gundalow knows, or @relrod?

**Comment by Andersson007** [comment]
@moonrail thanks for fixing this!
@felixfontein thanks for reviewing!
merged https://github.com/ansible-collections/community.general/pull/474 into main

@moonrail feel free to create an issue to upload new packages and update the tests. Then we could address the issue to a right person

**Pull request merged**